corehq/apps/hqwebapp/session_details_endpoint/tests.py

```diff
@@ -69,6 +71,59 @@ def test_session_details_view_updates_session(self):
         self.assertEqual(200, response.status_code)
         self.assertGreater(self.session.get_expiry_date(), expired_date)
 
+    @override_settings(DEBUG=True)
+    @flag_enabled('SECURE_SESSION_TIMEOUT')
+    def test_secure_sessions(self):
+
+        def _assertSecureSessionExpiry(previous_expiry, timeout_in_minutes):
+            self.assertEqual(True, self.client.session.get('secure_session'))
+            self.assertEqual(timeout_in_minutes, self.client.session.get('secure_session_timeout'))
+            self.assertNotEqual(previous_expiry, self.client.session.get_expiry_date())
+            expiring_in = (self.client.session.get_expiry_date() - datetime.datetime.utcnow()).seconds
+            self.assertGreater(expiring_in, timeout_in_minutes * 60 - 2)
+            self.assertLess(expiring_in, timeout_in_minutes * 60 + 2)
+
+        # Turn on secure sessions
+        self.domain.secure_sessions = True
+        self.domain.save()
+
+        # Sessions should now be secure and use timeout from settings
+        expired_date = self.session.get_expiry_date()
+        self.client.get(reverse('domain_homepage', args=[self.domain.name]))
+        _assertSecureSessionExpiry(expired_date, settings.SECURE_TIMEOUT)
+
+        # Turn on customized timeout
+        custom_timeout = 10
+        self.domain.secure_sessions_timeout = custom_timeout
+        self.domain.save()
+
+        # Request a domain-specific page so the domain-specific timeout kicks in
+        expired_date = self.session.get_expiry_date()
+        self.client.get(reverse('domain_homepage', args=[self.domain.name]))
+        _assertSecureSessionExpiry(expired_date, custom_timeout)
+
+        # Request a non-domain-specific page, which still uses the domain-specific timeout saved in the session
+        expired_date = self.session.get_expiry_date()
+        self.client.get(reverse('ping_session'))
+        _assertSecureSessionExpiry(expired_date, custom_timeout)
+
+        # Test the session details view itself
+        expired_date = self.session.get_expiry_date()
+        data = json.dumps({'sessionId': self.session_key, 'domain': self.domain.name})
+        response = Client().post(self.url, data, content_type="application/json")
+        self.assertEqual(200, response.status_code)
+        _assertSecureSessionExpiry(expired_date, custom_timeout)
+
+        # Going back to insecure sessions requires a logout to create a new session
+        self.domain.secure_sessions_timeout = None
+        self.domain.secure_sessions = False
+        self.domain.save()
+        self.client.logout()
+        self.client.login(username='bunkey', password='123')
+        self.client.get(reverse('domain_homepage', args=[self.domain.name]))
+        self.assertEqual(False, self.client.session.get('secure_session'))
+        self.assertEqual(settings.INACTIVITY_TIMEOUT, self.client.session.get('secure_session_timeout'))
+
```

Comment:
@millerdev and I also discussed this and settled on using the minimum timeout for all domains where the user is a member, because this matches the spirit of the secure_session flag (where you get the "most secure" experience or all of the domains you belong to). https://github.com/dimagi/commcare-hq/pull/27761/commits/07422d680bf9892a02dd1f96cd606cd3b1eacaf6 implements this. Definitely agree on test coverage, still working on that.

Reply:
015d8a524a6839d7e3fb273b1e9479a41de3db76 adds tests